### ISSUE TITLE
Label of ARIA progress bar not reported in browse mode if labelled by inner content

### ISSUE THREAD

**Issue by jcsteh** [nvaccess/nvda#16204]
### Steps to reproduce:
1. Open this test case:
    `data:text/html,<div role="progressbar" aria-labelledby="inner" aria-valuenow="10" aria-valuetext="10%"><span id="inner">World domination`
2. Press down arrow to read the single line in the document.

### Actual behavior:
NVDA says: "progress bar 10%"

### Expected behavior:
NVDA should say: "World domination  progress bar 10%"

#### NVDA installed/portable/running from source:
installed

#### NVDA version:
alpha-30378,b203f699 (2024.2.0.30378)

#### Windows version:
Windows 1 Version 23H2 (OS Build 22631.3155)

#### Name and version of other software in use when reproducing the issue:
Firefox Nightly 125.0a1 (2024-02-19) (64-bit)
Chrome Version 123.0.6312.0 (Official Build) canary (64-bit)

### Additional info
This occurs because of [this code](https://github.com/nvaccess/nvda/blob/6a43521f60eaafdea0521a68749a71e1328e4906/nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp#LL756) which means that we only force reporting of the name if the label is visible elsewhere. Normally, this makes perfect sense because we want to avoid double speaking of visible labels, since visible labels generally get rendered elsewhere in browse mode. Unfortunately, in the case of a progress bar, we never render inner content, so the label isn't *really* visible for our purposes.

The fix would be to consider a label invisible if it's labelling a progress bar (or slider) and the label is a descendant of said progress bar (or slider).

Arguably, this is a browser bug. The children of progress bar elements are pruned from the a11y tree, since they are presentational according to the spec. Given that, browsers probably shouldn't return them as the target of the labelledBy relation, since they're effectively "invisible" in the a11y tree. I'm happy to fix that in Firefox. However, I'm filing this here first because this affects both Firefox and Chrome and it may be less friction to fix it in NVDA. I know some Chromium folks feel it is reasonable to expose invisible relations and apparently Microsoft invested a lot of effort in making that work in Chromium, so you may get some push back from them in getting this "fixed" there.

**Comment by jcsteh** [comment]
> I know some Chromium folks feel it is reasonable to expose invisible relations and apparently Microsoft invested a lot of effort in making that work in Chromium, so you may get some push back from them in getting this "fixed" there.

CC @aleventhal for any thoughts on this point from the Chromium perspective.

**Comment by aleventhal** [comment]
@jcsteh are you sure this doesn't work as intended with NVDA and Chromium? In the tree I see a progressIndicator with the name "World domination". So that should work unless NVDA has some special detection for inner content or some other rule interfering, or?

The Microsoft work that went into exposing label and description relations when the target was invisible was only UIA, and I believe it never actually shipped. We do pick up the name and description when the target is invisible, as I believe all browsers do and the spec says to do.

Bottom line is we're fine with doing whatever the spec says, and I don't think we do anything different from what Firefox does, but maybe I'm misunderstanding the situation.

**Comment by jcsteh** [comment]
> @jcsteh are you sure this doesn't work as intended with NVDA and Chromium? In the tree I see a progressIndicator with the name "World domination".

The tree is correct in both browsers. This is about NVDA browse mode. In browse mode, NVDA has logic to try to avoid reporting the label twice if the label is visible.

What's happening here is that NVDA checks for a labelledBy relation, sees that the label is visible and thus doesn't report the label for the progress bar. But since the label isn't reachable from the a11y tree (it is pruned from the progress bar's children, as it should and must be), it never gets reported that way either.

> The Microsoft work that went into exposing label and description relations when the target was invisible was only UIA, and I believe it never actually shipped.

In this case, the target is visible, but it's not reachable. So, to speak in terms of the a11y tree:

1. The progress bar has no children because they are pruned. That means the label is not reachable when traversing the a11y tree.
2. However, labelledBy still returns the label, even though it isn't reachable.
3. I'm proposing that since the label isn't reachable, browsers shouldn't return it as a target of labelledBy.

**Comment by aleventhal** [comment]
Ah yeah, I was thinking we don't prune anything anymore because of all the times there was missing content. But I think progress bar may be an exception. I'll have to play around with it. But it sounds like we're saying the name should be exposed, but not the relation, because the relation object doesn't exist. I'm surprised if Chromium is even capable of exposing a target of a relation that isn't in the tree? ... end of day here, so fried, but I'll wait for your wisdom ...

**Comment by jcsteh** [comment]
I don't know the Chrome code particularly well, but it looks like the children of the progress bar are only pruned from the platform tree, not the core tree. However, they're pruned by way of [IsLeaf on the progress bar returning true](https://github.com/chromium/chromium/blob/fd8a8914ca0183f0add65ae55f04e287543c7d4a/ui/accessibility/ax_node.cc#LL1715), not IsIgnored on the children. The practical upshot is that the children aren't reachable via the progress bar (because the progress bar is a "leaf"), but they're still reachable via the relation (because they're not "ignored").

**Comment by jcsteh** [comment]
FWIW, Gecko has exactly the same bug for exactly the same reason.

**Comment by jcsteh** [comment]
The question we need to establish here is whether this is something we want to fix in browsers or NVDA. I tend to think it is more "correct" to fix this in browsers, but pragmatically, it's going to require fixes in multiple browsers, and it depends whether we agree that this relation shouldn't be exposed in this case.

**Comment by seanbudd** [comment]
Potentially related: #13307, #13912
